File: frontend/src/services/produitService.js
// src/services/produitService.js
import api from './api';

export const produitService = {
  // Récupérer tous les produits
  getAll: async () => {
    try {
      const response = await api.get('/produits');
      return response.data;
    } catch (error) {
      console.error('Erreur lors de la récupération des produits:', error);
      throw error;
    }
  },

  // Récupérer un produit par ID
  getById: async (id) => {
    try {
      const response = await api.get(`/produits/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Erreur lors de la récupération du produit ${id}:`, error);
      throw error;
    }
  },

  // Créer un nouveau produit
  create: async (produitData) => {
    try {
      const response = await api.post('/produits', produitData);
      return response.data;
    } catch (error) {
      console.error('Erreur lors de la création du produit:', error);
      throw error;
    }
  },

  // Mettre à jour un produit
  update: async (id, produitData) => {
    try {
      const response = await api.put(`/produits/${id}`, produitData);
      return response.data;
    } catch (error) {
      console.error(`Erreur lors de la mise à jour du produit ${id}:`, error);
      throw error;
    }
  },

  // Supprimer un produit
  delete: async (id) => {
    try {
      const response = await api.delete(`/produits/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Erreur lors de la suppression du produit ${id}:`, error);
      throw error;
    }
  },

  // Rechercher des produits
  search: async (query) => {
    try {
      const response = await api.get(`/produits/search?q=${encodeURIComponent(query)}`);
      return response.data;
    } catch (error) {
      console.error('Erreur lors de la recherche:', error);
      throw error;
    }
  },

  // Produits par catégorie
  getByCategorie: async (categorieId) => {
    try {
      const response = await api.get(`/produits/categorie/${categorieId}`);
      return response.data;
    } catch (error) {
      console.error(`Erreur lors de la récupération des produits de la catégorie ${categorieId}:`, error);
      throw error;
    }
  },

  // Produits en stock
  getEnStock: async () => {
    try {
      const response = await api.get('/produits/stock');
      return response.data;
    } catch (error) {
      console.error('Erreur lors de la récupération des produits en stock:', error);
      throw error;
    }
  }
};

File: backend/src/main/java/com/example/backend/service/ProduitService.java
package com.example.backend.service;

import com.example.backend.dto.request.ProduitRequest;
import com.example.backend.entity.Categorie;
import com.example.backend.entity.Produit;
import com.example.backend.repository.CategorieRepository;
import com.example.backend.repository.ProduitRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class ProduitService {

    private final ProduitRepository produitRepository;
    private final CategorieRepository categorieRepository;

    /**
     * Récupérer tous les produits
     */
    public List<Produit> getAllProduits() {
        return produitRepository.findAll();
    }

    /**
     * Récupérer un produit par ID
     */
    public Produit getProduitById(Long id) {
        return produitRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Produit non trouvé avec l'ID : " + id));
    }

    /**
     * Créer un nouveau produit
     */
    @Transactional
    public Produit createProduit(ProduitRequest request) {
        // Vérifier que la catégorie existe
        Categorie categorie = categorieRepository.findById(request.getCategorieId())
            .orElseThrow(() -> new RuntimeException("Catégorie non trouvée avec l'ID : " + request.getCategorieId()));

        // Créer le produit
        Produit produit = new Produit();
        produit.setNom(request.getNom());
        produit.setPrix(request.getPrix());
        produit.setQuantite(request.getQuantite());
        produit.setImage(request.getImage() != null ? request.getImage() : "https://via.placeholder.com/150");
        produit.setCategorie(categorie);
        produit.setAttributs(request.getAttributs());

        return produitRepository.save(produit);
    }

    /**
     * Mettre à jour un produit
     */
    @Transactional
    public Produit updateProduit(Long id, ProduitRequest request) {
        Produit produit = getProduitById(id);

        // Mettre à jour la catégorie si changée
        if (!produit.getCategorie().getId().equals(request.getCategorieId())) {
            Categorie categorie = categorieRepository.findById(request.getCategorieId())
                .orElseThrow(() -> new RuntimeException("Catégorie non trouvée avec l'ID : " + request.getCategorieId()));
            produit.setCategorie(categorie);
        }

        // Mettre à jour les autres champs
        produit.setNom(request.getNom());
        produit.setPrix(request.getPrix());
        produit.setQuantite(request.getQuantite());
        produit.setImage(request.getImage());
        produit.setAttributs(request.getAttributs());

        return produitRepository.save(produit);
    }

    /**
     * Supprimer un produit
     */
    @Transactional
    public void deleteProduit(Long id) {
        Produit produit = getProduitById(id);
        produitRepository.delete(produit);
    }

    /**
     * Rechercher des produits
     */
    public List<Produit> searchProduits(String search) {
        if (search == null || search.trim().isEmpty()) {
            return getAllProduits();
        }
        return produitRepository.searchProduits(search.trim());
    }

    /**
     * Récupérer les produits d'une catégorie
     */
    public List<Produit> getProduitsByCategorie(Long categorieId) {
        Categorie categorie = categorieRepository.findById(categorieId)
            .orElseThrow(() -> new RuntimeException("Catégorie non trouvée avec l'ID : " + categorieId));
        return produitRepository.findByCategorie(categorie);
    }

    /**
     * Récupérer les produits en stock
     */
    public List<Produit> getProduitsEnStock() {
        return produitRepository.findByQuantiteGreaterThan(0);
    }
}